### PR TITLE
- added possibility to write comment tags

### PR DESCRIPTION
**Pull request by brainbugfix**
I added the possibility to add the comments to the image during encoding as they come from decoding or added manually

### CODE REVIEW

File: lib/encoder.js [hunk 1]
```diff
@@ -535,6 +535,19 @@ function JPEGEncoder(quality) {
 				writeByte(std_ac_chrominance_values[p]);
 			}
 		}
+		
+		function writeCOM(comments)
+		{
+			if (typeof comments === "undefined") return;
```
Comment (patrickhulce) [inline lib/encoder.js:541]:
```suggestion
			if (typeof comments !== 'string') return;
```

perhaps?

**Review by patrickhulce — CHANGES_REQUESTED** [review]
thanks @brainbugfix! add a round-trip test for this?

types will also need to be updated

File: lib/encoder.js (hunk 1 above)
Comment (brainbugfix) [reply lib/encoder.js:541]: This will not work, I am afraid. The type is 'object' as it is an array.

Comment (brainbugfix) [reply lib/encoder.js:541]: Ah sorry, misunderstood your comment. I will change that tomorrow.

**Comment by patrickhulce** [comment]
thanks that change looks great, would you mind tackling...

> add a round-trip test for this?
types will also need to be updated

:D

**Comment by brainbugfix** [comment]
I am not sure if I understand what you want me to do. I can add a test, that should not be a big deal but what exactly do you mean with updating the types?

**Comment by patrickhulce** [comment]
Oh sorry, I mean update the typescript definitions in `index.d.ts` at https://github.com/jpeg-js/jpeg-js/blob/417e8e2baff2d1e546fdf68c492fba88205d3fb5/index.d.ts#L13

It's already a bit out of date but just changing the first encode parameter to `RawImageData<BufferLike> & {comments?: string[]}` should work?

**Comment by brainbugfix** [comment]
Did what you asked me for. Can you please double-check the TypeScript because I've never used this before.

File: index.d.ts [hunk 2]
```diff
@@ -2,6 +2,7 @@ export interface RawImageData<T> {
   width: number;
   height: number;
   data: T;
+  comments?: string[];
```
Comment (patrickhulce) [inline index.d.ts:5]:
```suggestion
```

File: index.d.ts [hunk 3]
```diff
@@ -10,7 +11,7 @@ type UintArrRet = RawImageData<Uint8Array>;
 type ImageData = BufferRet | UintArrRet;
 type BufferLike = Buffer | Uint8Array | ArrayLike<number> | Iterable<number> | ArrayBuffer;
 
-export declare function encode(imgData: RawImageData<BufferLike>, quality?: number): BufferRet;
+export declare function encode(imgData: RawImageData<BufferLike>, quality?: number): BufferRet & {comments?: string[]};
```
Comment (patrickhulce) [inline index.d.ts:14]:
```suggestion
export declare function encode(imgData: RawImageData<BufferLike> & {comments?: string[]}, quality?: number): BufferRet
```

File: test/index.js [hunk 4]
```diff
@@ -179,6 +179,31 @@ it('should be able to create a JPEG from an array', function () {
...
+    data: frameData,
+    width: width,
+    height: height,
+    comments: comments,
+  };
+  var jpegImageData = jpeg.encode(rawImageData, 50); 
+  expect(jpegImageData.width).toEqual(width);
+  expect(jpegImageData.height).toEqual(height);
+  var expected = fixture('redbox_comment.jpg');
+  expect(jpegImageData.data).toEqual(expected);
```
Comment (patrickhulce) [inline test/index.js:204]:
```suggestion
  expect(jpegImageData.data).toEqual(expected);
  expect(jpeg.decode(jpegImageData.data).comments).toEqual(['First comment', 'second comment']);
```

this is what I meant by a round-trip test :)

**Review by patrickhulce — APPROVED** [review]
thanks @brainbugfix this is great!

File: lib/encoder.js [hunk 5]
```diff
@@ -782,7 +797,8 @@ function encode(imgData, qu) {
   return {
     data: data,
     width: imgData.width,
-    height: imgData.height
+    height: imgData.height,
+	comments: imgData.comments
```
Comment (patrickhulce) [inline lib/encoder.js:801]:
```suggestion
```

sorry I missed this, I don't think we should return here. is there a reason you need to?

Comment (brainbugfix) [reply lib/encoder.js:801]: There is no specific reason. It was more to have the same output in both functions.

**Comment by patrickhulce** [comment]
if all these suggestions are ok with you @brainbugfix I'm ready to merge! :)

File: test/index.js [hunk 6]
```diff
@@ -179,6 +179,32 @@ it('should be able to create a JPEG from an array', function () {
...
+  expect(jpeg.decode(jpegImageData.data).comments).toEqual(['First comment', 'second comment']);
```
Comment (patrickhulce) [inline test/index.js:205]:
```suggestion
  expect(jpeg.decode(jpegImageData.data).comments).toEqual(['First comment', 'Second comment']);
```

**Comment by patrickhulce** [comment]
thanks again :)

**Pull request merged**